### ISSUE TITLE
Python 3.12 and numpy

### ISSUE THREAD

**Issue by tuhe** [dtudk/dtumathtools#7]
Hi!

We might as well start thinking about python 3.12. I tried playing around with 3.12.03b for a bit and I don't think it is compatible with numpy <=0.24 (due to distutils being removed in 3.12) -- so I think it would be a good idea to think about bumping the recommended numpy version (which is at the moment pinned at <0.24).

**Comment by Chrillebon** [comment]
I assume that you refer to numpy <= **1**.24.

The version set for numpy was set to match the version of [spb](https://github.com/Davide-sd/sympy-plot-backends/blob/master/requirements.txt).

But agreed, support for 3.12 is a must going forward!

**Comment by Chrillebon** [comment]
I am working on upgrading to SPB ~=2.4, which should be out in a release soon. 

Wrt. `quiver` vs `plot_vector`, as noted [here](https://sympy-plot-backends.readthedocs.io/en/latest/modules/vectors.html#spb.vectors.plot_vector), the `plot_vector` only handles 2D or 3D vector fields. Because of this, the functionality behind `quiver` (plotting single vectors) is still needed and should be updated if needed.

However, as stated by @tuhe, Python 3.12 does not work currently because of distutils through numpy (tested in 3.12.0b4). It seems that [updates will be coming](https://github.com/numpy/numpy/issues/23808#issuecomment-1599686864), but I will be supporting 3.8-3.11 for now.

**Comment by horsten** [comment]
I just hacked a version of dtumathtools that works with 3.12 for the purposes of the Programming course. All I did was bump up everything to the latest version. I don't know how full your test coverage is, but apart from the plotting features, all the tests pass, so maybe it was an idea to split the package into two, separating the volatile GUI stuff out in a separate dtumathplot package?

![image](https://github.com/dtudk/dtumathtools/assets/3442409/5001f687-24cf-4d0d-b474-6ce89f0ecd58)

```
PS C:\Users\th\src\python\dtumathtools> pip list
Package                   Version      Editable project location
------------------------- ------------ ------------------------------------
anyio                     4.0.0
appdirs                   1.4.4
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.0
async-lru                 2.0.4
attrs                     23.1.0
Babel                     2.13.0
backcall                  0.2.0
beautifulsoup4            4.12.2
bidict                    0.22.1
bleach                    6.1.0
blinker                   1.6.3
build                     1.0.3
certifi                   2023.7.22
cffi                      1.16.0
charset-normalizer        3.3.0
click                     8.1.7
colorama                  0.4.6
comm                      0.1.4
contourpy                 1.1.1
coverage                  7.3.2
cp                        2023.0       C:\Users\th\src\python\02002students
cycler                    0.12.1
debugpy                   1.8.0
decorator                 5.1.1
defusedxml                0.7.1
diskcache                 5.6.3
dtumathtools              1.1.0+th1
executing                 2.0.0
fastjsonschema            2.18.1
Flask                     3.0.0
Flask-SocketIO            5.3.6
fonttools                 4.43.1
fqdn                      1.5.1
h11                       0.14.0
idna                      3.4
importnb                  2023.1.7
iniconfig                 2.0.0
ipykernel                 6.25.2
ipython                   8.16.1
ipython-genutils          0.2.0
ipywidgets                8.1.1
isoduration               20.11.0
itsdangerous              2.1.2
jedi                      0.19.1
Jinja2                    3.1.2
json5                     0.9.14
jsonpointer               2.4
jsonschema                4.19.1
jsonschema-specifications 2023.7.1
jupyter                   1.0.0
jupyter_client            8.4.0
jupyter-console           6.6.3
jupyter_core              5.4.0
jupyter-events            0.7.0
jupyter-lsp               2.2.0
jupyter_server            2.7.3
jupyter_server_terminals  0.4.4
jupyterlab                4.0.7
jupyterlab-pygments       0.2.2
jupyterlab_server         2.25.0
jupyterlab-widgets        3.0.9
kiwisolver                1.4.5
MarkupSafe                2.1.3
matplotlib                3.8.0
matplotlib-inline         0.1.6
mergedeep                 1.3.4
mistune                   3.0.2
mpmath                    1.3.0
nbclient                  0.8.0
nbconvert                 7.9.2
nbformat                  5.9.2
nest-asyncio              1.5.8
notebook                  7.0.4
notebook_shim             0.2.3
numpy                     1.26.0
overrides                 7.4.0
packaging                 23.2
pandas                    2.1.1
pandocfilters             1.5.0
parso                     0.8.3
pickleshare               0.7.5
Pillow                    10.0.1
pip                       23.2.1
platformdirs              3.11.0
pluggy                    1.3.0
prometheus-client         0.17.1
prompt-toolkit            3.0.39
psutil                    5.9.5
pure-eval                 0.2.2
pycparser                 2.21
pyfiglet                  0.8.post1
Pygments                  2.16.1
pyparsing                 3.1.1
pyproject_hooks           1.0.0
pytest                    7.4.2
python-dateutil           2.8.2
python-engineio           4.7.1
python-json-logger        2.0.7
python-socketio           5.9.0
pytz                      2023.3.post1
pywin32                   306
pywinpty                  2.0.12
PyYAML                    6.0.1
pyzmq                     25.1.1
qtconsole                 5.4.4
QtPy                      2.4.0
referencing               0.30.2
requests                  2.31.0
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rpds-py                   0.10.6
Send2Trash                1.8.2
setuptools                68.2.2
simple-websocket          1.0.0
six                       1.16.0
sniffio                   1.3.0
soupsieve                 2.5
stack-data                0.6.3
sympy                     1.12
sympy-plot-backends       3.0.0
tabulate                  0.9.0
terminado                 0.17.1
tinycss2                  1.2.1
tornado                   6.3.3
tqdm                      4.66.1
traitlets                 5.11.2
types-python-dateutil     2.8.19.14
tzdata                    2023.3
unitgrade                 1.0.0.7
uri-template              1.3.0
urllib3                   2.0.6
watchdog                  3.0.0
wcwidth                   0.2.8
webcolors                 1.13
webencodings              0.5.1
websocket-client          1.6.4
Werkzeug                  3.0.0
widgetsnbextension        4.0.9
wsproto                   1.2.0
```

**Comment by zerothi** [comment]
@horsten Thanks for this.
We will first bump a new release in the coming semester. We don't want to push students into upgrading python, or have students running different versions of everything (it is tedious to debug since there may be additional unknowns).

But, this is great for the coming release.
Lastly, we will not split the package into two. The whole point is that the dtumathtools is a *complete* package for the students, with all necessary dependencies, including plotting facilities.

**Comment by Chrillebon** [comment]
Agree with @zerothi in that this is "the one package that students need". 

With respect to the test coverage, all functionality (lines of code) that was presented throughout the 21/22 demos for the course are being checked in the current tests.

**Comment by tuhe** [comment]
Hi! For the spring semester, it would be ideal to support 3.12 in intro to programming since many will install it by default. Do you plan to update the version on pypi to lift the version pin (<= 0.24) on numpy, or is 3.12/numpy still causing issues with your other packages?

**Comment by Chrillebon** [comment]
Yes! It would be ideal to support 3.12, and have an update out on pypi before the new semester! Think we just need to find versions that work (eg. like [this](https://github.com/dtudk/dtumathtools/pull/19#issue-1938954806)). Will start off a branch with the spring semester to try to find package versions that work. Maybe @zerothi can note/remember what else we wanted for the spring semester (eg. version numbering 2024.1)

**Comment by zerothi** [comment]
This has been fixed in #24, closing
